File: ios/NativeEditorViewRegistry.swift
import UIKit

enum NativeEditorDestroyReservationResult: Equatable {
    case reserved
    case alreadyInProgress
    case unavailable
}

final class NativeEditorViewRegistry {
    static let shared = NativeEditorViewRegistry()

    private final class RegisteredView {
        weak var view: NativeEditorExpoView?
        let order: UInt64

        init(view: NativeEditorExpoView, order: UInt64) {
            self.view = view
            self.order = order
        }
    }

    private var viewsByEditorId: [UInt64: [RegisteredView]] = [:]
    private var nextRegistrationOrder: UInt64 = 0
    private var activeEditorIds: Set<UInt64> = []
    private var destroyingEditorIds: Set<UInt64> = []
    var onFinalizeDestroyForTesting: ((UInt64) throws -> Void)?

    private init() {}

    func markEditorCreated(editorId: UInt64) {
        guard editorId != 0 else { return }
        _ = performOnMain {
            activeEditorIds.insert(editorId)
        }
    }

    func rebaseAfterRemoteCommit(editorId: UInt64) {
        guard editorId != 0 else { return }
        DispatchQueue.main.async { [weak self] in
            self?.applyRemoteCommitRefresh(editorId: editorId)
        }
    }

    func applyRemoteCommitRefresh(editorId: UInt64) {
        dispatchPrecondition(condition: .onQueue(.main))
        liveRegisteredViews(editorId: editorId).forEach { $0.view?.applyRemoteCommitRefresh() }
    }

    func isDestroyed(editorId: UInt64) -> Bool {
        guard editorId != 0 else { return false }
        return performOnMain {
            if destroyingEditorIds.contains(editorId) { return true }
            if activeEditorIds.contains(editorId) {
                return false
            }
            return EditorV2Registry.adapter(forLegacyId: editorId) == nil
        }
    }

    @discardableResult
    func register(editorId: UInt64, view: NativeEditorExpoView) -> Bool {
        guard editorId != 0 else { return false }
        return performOnMain {
            guard !destroyingEditorIds.contains(editorId) else { return false }
            if !activeEditorIds.contains(editorId) {
                guard EditorV2Registry.adapter(forLegacyId: editorId) != nil else {
                    return false
                }
                activeEditorIds.insert(editorId)
            }
            var views = liveRegisteredViews(editorId: editorId)
            guard !views.contains(where: { $0.view === view }) else { return true }
            nextRegistrationOrder &+= 1
            views.append(RegisteredView(view: view, order: nextRegistrationOrder))
            viewsByEditorId[editorId] = views
            return true
        }
    }

    func unregister(editorId: UInt64, view: NativeEditorExpoView) {
        guard editorId != 0 else { return }
        performOnMain {
            let views = liveRegisteredViews(editorId: editorId).filter { $0.view !== view }
            if views.isEmpty {
                viewsByEditorId.removeValue(forKey: editorId)
            } else {
                viewsByEditorId[editorId] = views
            }
        }
    }

    func nativeOwnerReleased(editorId: UInt64, by view: NativeEditorExpoView) {
        guard editorId != 0 else { return }
        performOnMain {
            guard !destroyingEditorIds.contains(editorId) else { return }
            let survivor = liveRegisteredViews(editorId: editorId)
                .reversed()
                .compactMap(\.view)
                .first { $0 !== view && $0.window != nil }
            survivor?.claimNativeOwnershipAndCatchUp(editorId: editorId)
        }
    }

    func invalidateDestroyedEditor(editorId: UInt64) {
        guard reserveDestroy(editorId: editorId) else { return }
        finalizeDestroy(editorId: editorId)
    }

    /// Install the reversible, in-flight destroy gate before crossing the
    /// FFI boundary. The active view state remains intact until finalization
    /// so a retryable destroy can roll back without reconstructing bindings.
    func acquireDestroyReservation(editorId: UInt64) -> NativeEditorDestroyReservationResult {
        guard editorId != 0 else { return .unavailable }
        return performOnMain {
            guard !destroyingEditorIds.contains(editorId) else { return .alreadyInProgress }
            guard activeEditorIds.contains(editorId)
                || EditorV2Registry.adapter(forLegacyId: editorId) != nil
            else {
                return .unavailable
            }
            destroyingEditorIds.insert(editorId)
            return .reserved
        }
    }

    @discardableResult
    func reserveDestroy(editorId: UInt64) -> Bool {
        acquireDestroyReservation(editorId: editorId) == .reserved
    }

    func rollbackDestroy(editorId: UInt64) {
        guard editorId != 0 else { return }
        _ = performOnMain {
            destroyingEditorIds.remove(editorId)
        }
    }

    func isDestroyReserved(editorId: UInt64) -> Bool {
        guard editorId != 0 else { return false }
        return performOnMain {
            destroyingEditorIds.contains(editorId)
        }
    }

    func finalizeDestroy(editorId: UInt64) {
        guard editorId != 0 else { return }
        performOnMain {
            guard destroyingEditorIds.contains(editorId) else { return }
            invokeDestroyTestingHook(onFinalizeDestroyForTesting, editorId: editorId)
            activeEditorIds.remove(editorId)
            let views = viewsByEditorId.removeValue(forKey: editorId)?.compactMap(\.view) ?? []
            views.forEach { $0.handleEditorDestroyed(editorId) }
            destroyingEditorIds.remove(editorId)
        }
    }

    func destroy(editorId: UInt64, operation: () -> Void) {
        guard reserveDestroy(editorId: editorId) else { return }
        operation()
        finalizeDestroy(editorId: editorId)
    }

    func prepareForCommandJSON(editorId: UInt64) -> String {
        let prepare = { () -> String in
            if self.destroyingEditorIds.contains(editorId) {
                return Self.commandPreparationJSON(ready: false, blockedReason: "destroying")
            }
            if !self.activeEditorIds.contains(editorId),
               EditorV2Registry.adapter(forLegacyId: editorId) == nil {
                return Self.commandPreparationJSON(ready: false, blockedReason: "destroyed")
            }
            let views = self.liveRegisteredViews(editorId: editorId).compactMap(\.view)
            guard let view = views.reversed().first(where: {
                $0.ownsNativeBinding(editorId: editorId)
            }) ?? views.last else {
                self.viewsByEditorId.removeValue(forKey: editorId)
                return Self.commandPreparationJSON(ready: true)
            }
            return view.prepareForEditorCommandJSON()
        }

        return performOnMain(prepare)
    }

    private func liveRegisteredViews(editorId: UInt64) -> [RegisteredView] {
        let views = viewsByEditorId[editorId]?.filter { $0.view != nil } ?? []
        if views.isEmpty {
            viewsByEditorId.removeValue(forKey: editorId)
        } else {
            viewsByEditorId[editorId] = views
        }
        return views.sorted { $0.order < $1.order }
    }

    static func commandPreparationJSON(
        ready: Bool,
        updateJSON: String? = nil,
        blockedReason: String? = nil
    ) -> String {
        var payload: [String: Any] = ["ready": ready]
        if let updateJSON {
            payload["updateJSON"] = updateJSON
        }
        if let blockedReason {
            payload["blockedReason"] = blockedReason
        }
        guard let data = try? JSONSerialization.data(withJSONObject: payload),
              let json = String(data: data, encoding: .utf8)
        else {
            if let blockedReason {
                return ready
                    ? "{\"ready\":true,\"blockedReason\":\"\(blockedReason)\"}"
                    : "{\"ready\":false,\"blockedReason\":\"\(blockedReason)\"}"
            }
            return ready ? "{\"ready\":true}" : "{\"ready\":false}"
        }
        return json
    }

    private func performOnMain<T>(_ work: () -> T) -> T {
        if Thread.isMainThread {
            return work()
        }
        return DispatchQueue.main.sync(execute: work)
    }
}

File: android/src/main/java/com/apollohg/editor/NativeEditorViewRegistry.kt
package com.apollohg.editor

import android.os.Handler
import android.os.Looper
import java.lang.ref.WeakReference
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicReference
import org.json.JSONObject

private const val DESTROY_INVALIDATION_AWAIT_TIMEOUT_MS = 250L

internal enum class NativeEditorDestroyReservationResult {
    RESERVED,
    ALREADY_IN_PROGRESS,
    UNAVAILABLE
}

private class WeakNativeEditorExpoView private constructor(
    val view: WeakReference<NativeEditorExpoView?>
) {
    constructor(view: NativeEditorExpoView) : this(WeakReference(view))

    companion object {
        fun cleared(): WeakNativeEditorExpoView =
            WeakNativeEditorExpoView(WeakReference<NativeEditorExpoView?>(null))
    }
}

internal object NativeEditorViewRegistry {
    private data class CommandPreparationSnapshot(
        val view: NativeEditorExpoView?,
        val isDetached: Boolean,
        val isDestroyed: Boolean
    )

    private val liveEditorIds = mutableSetOf<Long>()
    private val viewsByEditorId = mutableMapOf<Long, MutableList<WeakNativeEditorExpoView>>()
    private val inputViewsByEditorId =
        mutableMapOf<Long, MutableList<WeakReference<EditorEditText>>>()
    private val detachedEditorOwnersByEditorId = mutableMapOf<Long, WeakNativeEditorExpoView>()
    private val destroyingEditorIds = mutableSetOf<Long>()
    private val destroyReservationWasLive = mutableMapOf<Long, Boolean>()
    private val mainHandler = Handler(Looper.getMainLooper())

    @Volatile
    internal var onFinalizeDestroyForTesting: ((Long) -> Unit)? = null

    @Synchronized
    fun markEditorCreated(editorId: Long) {
        if (editorId == 0L) return
        liveEditorIds.add(editorId)
        destroyingEditorIds.remove(editorId)
        destroyReservationWasLive.remove(editorId)
    }

    @Synchronized
    fun register(editorId: Long, view: NativeEditorExpoView): Boolean {
        if (editorId == 0L) return false
        if (destroyingEditorIds.contains(editorId)) return false
        val isKnownDetachedOwner = detachedEditorOwnersByEditorId[editorId]?.view?.get() === view
        if (
            !liveEditorIds.contains(editorId) &&
            !isKnownDetachedOwner &&
            !rustEditorExists(editorId)
        ) {
            return false
        }
        val views = viewsByEditorId.getOrPut(editorId) { mutableListOf() }
        views.removeAll { it.view.get() == null || it.view.get() === view }
        views += WeakNativeEditorExpoView(view)
        detachedEditorOwnersByEditorId.remove(editorId)
        return true
    }

    @Synchronized
    fun unregister(
        editorId: Long,
        view: NativeEditorExpoView,
        blockCommandsUntilRegistered: Boolean = false
    ) {
        if (editorId == 0L) return
        val registeredViews = viewsByEditorId[editorId]
        val wasRegistered = registeredViews?.any { it.view.get() === view } == true
        registeredViews?.removeAll { it.view.get() == null || it.view.get() === view }
        if (registeredViews?.isEmpty() == true) viewsByEditorId.remove(editorId)
        if (blockCommandsUntilRegistered) {
            detachedEditorOwnersByEditorId[editorId] = WeakNativeEditorExpoView(view)
        } else {
            val detachedOwner = detachedEditorOwnersByEditorId[editorId]?.view?.get()
            if (wasRegistered || detachedOwner === view) {
                detachedEditorOwnersByEditorId.remove(editorId)
            }
        }
    }

    @Synchronized
    fun isDestroyed(editorId: Long): Boolean = destroyingEditorIds.contains(editorId)

    @Synchronized
    internal fun retainedDestroyedIdCountForTests(): Int = destroyingEditorIds.size

    @Synchronized
    internal fun forceDetachedOwnerClearedForTesting(editorId: Long) {
        detachedEditorOwnersByEditorId[editorId] = WeakNativeEditorExpoView.cleared()
    }

    @Synchronized
    internal fun forceRegisteredViewsClearedForTesting(editorId: Long) {
        viewsByEditorId[editorId] = mutableListOf(WeakNativeEditorExpoView.cleared())
    }

    @Synchronized
    internal fun boundViewReferenceCountForTests(editorId: Long): Int =
        viewsByEditorId[editorId]?.size ?: 0

    @Synchronized
    private fun liveViewsFor(editorId: Long): List<NativeEditorExpoView> =
        viewsByEditorId[editorId]?.mapNotNull { it.view.get() }.orEmpty()

    fun rebaseAfterRemoteCommit(handle: String) {
        val viewToken = EditorV2Registry.viewTokenForHandle(handle) ?: return
        liveViewsFor(viewToken).forEach { view ->
            if (view.markRemoteCommitRebaseScheduled(viewToken)) {
                mainHandler.post { view.applyRemoteCommitRefresh(viewToken) }
            }
        }
    }

    @Synchronized
    fun registerInputView(editorId: Long, view: EditorEditText) {
        if (editorId == 0L || destroyingEditorIds.contains(editorId)) return
        if (!liveEditorIds.contains(editorId) && !rustEditorExists(editorId)) return
        val views = inputViewsByEditorId.getOrPut(editorId) { mutableListOf() }
        views.removeAll { it.get() == null || it.get() === view }
        views += WeakReference(view)
    }

    @Synchronized
    fun acquireDestroyReservation(editorId: Long): NativeEditorDestroyReservationResult {
        if (editorId == 0L) return NativeEditorDestroyReservationResult.UNAVAILABLE
        if (destroyingEditorIds.contains(editorId)) {
            return NativeEditorDestroyReservationResult.ALREADY_IN_PROGRESS
        }
        if (!liveEditorIds.contains(editorId) && !rustEditorExists(editorId)) {
            return NativeEditorDestroyReservationResult.UNAVAILABLE
        }
        destroyingEditorIds.add(editorId)
        destroyReservationWasLive[editorId] = liveEditorIds.remove(editorId)
        return NativeEditorDestroyReservationResult.RESERVED
    }

    fun beginDestroy(editorId: Long): Boolean =
        acquireDestroyReservation(editorId) == NativeEditorDestroyReservationResult.RESERVED

    /** Undo an in-flight reservation after an FFI result that is retryable or malformed. */
    @Synchronized
    fun rollbackDestroy(editorId: Long) {
        if (!destroyingEditorIds.remove(editorId)) return
        if (destroyReservationWasLive.remove(editorId) == true) {
            liveEditorIds.add(editorId)
        }
    }

    fun invalidateDestroyedEditor(editorId: Long) {
        if (!beginDestroy(editorId)) return
        finalizeDestroy(editorId)
    }

    /**
     * Finalize a reserved destroy. Off-main callers may time out while the
     * main-thread invalidation is still queued, so completion is reported only
     * after the reservation itself has been released.
     */
    fun finalizeDestroy(editorId: Long, onCompleted: (() -> Unit)? = null) {
        val affectedViews = synchronized(this) {
            if (!destroyingEditorIds.contains(editorId)) return@synchronized null
            val views = listOfNotNull(
                *viewsByEditorId.remove(editorId)
                    .orEmpty()
                    .mapNotNull { it.view.get() }
                    .toTypedArray(),
                detachedEditorOwnersByEditorId.remove(editorId)?.view?.get()
            ).distinct()
            val inputViews = inputViewsByEditorId.remove(editorId)
                .orEmpty()
                .mapNotNull { it.get() }
                .distinct()
            views to inputViews
        }
        if (affectedViews == null) {
            onCompleted?.invoke()
            return
        }
        fun releaseReservation() {
            synchronized(this) {
                destroyingEditorIds.remove(editorId)
                destroyReservationWasLive.remove(editorId)
            }
            onCompleted?.invoke()
        }
        invokeDestroyTestingHook(onFinalizeDestroyForTesting, editorId)
        if (affectedViews.first.isEmpty() && affectedViews.second.isEmpty()) {
            releaseReservation()
            return
        }
        val invalidate = Runnable {
            try {
                affectedViews.first.forEach { view ->
                    view.handleEditorDestroyed(editorId)
                }
                affectedViews.second.forEach { view ->
                    view.handleEditorDestroyedFromRegistry(editorId)
                }
            } finally {
                releaseReservation()
            }
        }
        if (Looper.myLooper() == Looper.getMainLooper()) {
            invalidate.run()
        } else {
            val latch = CountDownLatch(1)
            val posted = mainHandler.post {
                try {
                    invalidate.run()
                } finally {
                    latch.countDown()
                }
            }
            if (!posted) {
                releaseReservation()
                return
            }
            try {
                latch.await(DESTROY_INVALIDATION_AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS)
            } catch (_: InterruptedException) {
                Thread.currentThread().interrupt()
            } catch (_: Throwable) {
                return
            }
        }
    }

    fun prepareForCommandJSON(editorId: Long): String {
        val prepare = {
            val snapshot = synchronized(this) {
                val isDestroyed = destroyingEditorIds.contains(editorId)
                if (isDestroyed) {
                    return@synchronized CommandPreparationSnapshot(
                        view = null,
                        isDetached = false,
                        isDestroyed = true
                    )
                }
                val registeredViews = viewsByEditorId[editorId]
                registeredViews?.removeAll { it.view.get() == null }
                val candidate = registeredViews?.firstNotNullOfOrNull { it.view.get() }
                if (registeredViews?.isEmpty() == true) {
                    viewsByEditorId.remove(editorId)
                }
                val detachedOwner = detachedEditorOwnersByEditorId[editorId]?.view?.get()
                val isDetached = if (detachedOwner == null) {
                    detachedEditorOwnersByEditorId.remove(editorId)
                    false
                } else {
                    true
                }
                val missingFromRust =
                    !liveEditorIds.contains(editorId) &&
                        candidate == null &&
                        !isDetached &&
                        !rustEditorExists(editorId)
                CommandPreparationSnapshot(
                    view = candidate,
                    isDetached = isDetached,
                    isDestroyed = missingFromRust
                )
            }
            snapshot.view?.prepareForEditorCommandJSON()
                ?: commandPreparationJSON(
                    ready = !snapshot.isDetached && !snapshot.isDestroyed,
                    blockedReason = if (snapshot.isDestroyed) {
                        "destroyed"
                    } else if (snapshot.isDetached) {
                        "detached"
                    } else {
                        null
                    }
                )
        }

        if (Looper.myLooper() == Looper.getMainLooper()) {
            return prepare()
        }

        val result =
            AtomicReference(commandPreparationJSON(ready = false, blockedReason = "unknown"))
        val state = AtomicInteger(PREFLIGHT_STATE_QUEUED)
        val latch = CountDownLatch(1)
        if (!mainHandler.post {
                try {
                    if (state.compareAndSet(PREFLIGHT_STATE_QUEUED, PREFLIGHT_STATE_RUNNING)) {
                        result.set(prepare())
                        state.set(PREFLIGHT_STATE_DONE)
                    }
                } finally {
                    latch.countDown()
                }
            }
        ) {
            return commandPreparationJSON(ready = false, blockedReason = "unknown")
        }
        try {
            if (!latch.await(DESTROY_INVALIDATION_AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS)) {
                if (state.compareAndSet(PREFLIGHT_STATE_QUEUED, PREFLIGHT_STATE_CANCELLED)) {
                    return commandPreparationJSON(ready = false, blockedReason = "unknown")
                }
                if (state.get() == PREFLIGHT_STATE_RUNNING) {
                    latch.await()
                    return result.get()
                }
                return commandPreparationJSON(ready = false, blockedReason = "unknown")
            }
        } catch (_: InterruptedException) {
            var interrupted = true
            if (state.compareAndSet(PREFLIGHT_STATE_QUEUED, PREFLIGHT_STATE_CANCELLED)) {
                Thread.currentThread().interrupt()
                return commandPreparationJSON(ready = false, blockedReason = "unknown")
            }
            while (state.get() == PREFLIGHT_STATE_RUNNING) {
                try {
                    latch.await()
                    break
                } catch (_: InterruptedException) {
                    interrupted = true
                }
            }
            if (interrupted) {
                Thread.currentThread().interrupt()
            }
            return if (state.get() == PREFLIGHT_STATE_DONE) {
                result.get()
            } else {
                commandPreparationJSON(ready = false, blockedReason = "unknown")
            }
        }
        return result.get()
    }

    private fun rustEditorExists(viewToken: Long): Boolean =
        EditorV2Registry.adapterForViewToken(viewToken) != null

    fun commandPreparationJSON(
        ready: Boolean,
        updateJSON: String? = null,
        blockedReason: String? = null
    ): String = JSONObject().apply {
        put("ready", ready)
        if (updateJSON != null) {
            put("updateJSON", updateJSON)
        }
        if (!ready && blockedReason != null) {
            put("blockedReason", blockedReason)
        }
    }.toString()

    private const val PREFLIGHT_STATE_QUEUED = 0
    private const val PREFLIGHT_STATE_RUNNING = 1
    private const val PREFLIGHT_STATE_CANCELLED = 2
    private const val PREFLIGHT_STATE_DONE = 3
}
